File: proposed/local_optimization.py
from utilities.point import fermat_point
from utilities.tree import Tree

from proposed.weiszfeld import weiszfeld


def local_optimization(T: Tree, S: list):
    """
    Thuật toán 6: Local Optimization
    Tìm kiếm cục bộ nhằm tối ưu vị trí các điểm Steiner với cây đã cho:
    - Giả sử các hàng xóm của điểm không đổi trong quá trình
    - Nếu điểm có 3 hàng xóm, đặt vị trí dlaf điểm Fermat của tam giác
    - Nếu có nhiều hàng xóm, dùng giải thuật Weiszfeld
    - Lặp lại quá trình trên tới khi tổng trọng số không giảm nữa
    """

    length = float('inf')

    while T.l < length - 1e-9:
        length = T.l

        for s in S:
            if T.degree(s) == 3:
                new_s = fermat_point(T.adjacency_list[s])
            else:
                new_s = weiszfeld(s, T.adjacency_list[s]).axes
            s.set_axes(new_s)

    # Loại bỏ các điểm Steiner trùng với hàng xóm
    removed_nodes = []
    for s in S:
        for p in T.adjacency_list[s]:
            if s.equals(p):
                removed_nodes.append(s)
                T.remove_edge(s, p)
                for q in T.adjacency_list[s]:
                    T.add_edge(p, q)
                T.remove_node(s)
                break

    temp_S = []
    length = len(removed_nodes)
    it = 0
    for s in S:
        if it < length and s is removed_nodes[it]:
            it += 1
        else:
            temp_S.append(s)
    return temp_S

File: cpp1/local_optimization.cpp
#include "local_optimization.hpp"
#include <cmath>
#include <algorithm>

std::vector<Point*> local_optimization(Tree& T, std::vector<Point*>& S) {
    double length = std::numeric_limits<double>::infinity();

    while (T.l() < length - 1e-9) {
        length = T.l();

        for (auto s : S) {
            if (T.degree(s) == 3) {
                std::vector<Point*> neighbors(T.adjacency_list[s].begin(), T.adjacency_list[s].end());
                std::vector<Point> neighbor_points;
                for (auto p : neighbors) {
                    neighbor_points.push_back(*p);
                }
                std::vector<double> new_s = fermat_point(neighbor_points);
                s->set_axes(new_s);
            } else {
                std::vector<Point> neighbors;
                for (auto p : T.adjacency_list[s]) {
                    neighbors.push_back(*p);
                }
                Point new_s = weiszfeld(*s, neighbors);
                s->set_axes(new_s.axes);
            }
        }
    }

    std::vector<Point*> removed_nodes;
    for (auto s : S) {
        for (auto p : T.adjacency_list[s]) {
            if (s->equals(*p)) {
                removed_nodes.push_back(s);
                T.remove_edge(s, p);
                for (auto q : T.adjacency_list[s]) {
                    T.add_edge(p, q);
                }
                T.remove_node(s);
                break;
            }
        }
    }

    std::vector<Point*> temp_S;
    auto it = removed_nodes.begin();
    for (auto s : S) {
        if (it != removed_nodes.end() && s == *it) {
            ++it;
        } else {
            temp_S.push_back(s);
        }
    }
    return temp_S;
}